### ISSUE TITLE
Solution for dynamic partitions?

### ISSUE THREAD

**Issue by kheide** [balderdashy/sails-postgresql#162]
Is there any way in the adapter to this date that makes it possible to dynamically search a specific partition in a postgres db? That would pretty much solve the multi-tenancy issue atleast for me.

**Comment by tjwebb** [comment]
Everyone has a different definition for what they think "multi-tenancy" means. Can you be more specific about what problem you're trying to solve?

**Comment by kheide** [comment]
I would love to! 

Lets take two platforms as examples. Instagram and Salesforce. Instagram would not be multi-tenant, but Salesforce would. 

I am trying to build a platform where I can house multiple customers with their separate set of data, much like Salesforce and Salesforce actually does this on their force1 platform by using partitioning. This is what I am trying to do with Sails as well. I am not hell-bent on using partitioning as the preferred solution would be to put each customer in their own db and dynamically route the requests to the correct db based on for example the sub-domain or any other parameter in req.session. I've seen a few implementations off this which has not made it into Sails.js so thats why I wondered if it was possible to do it with partitioning and postgresql in the current state of Sails. 

Did I manage to get my use case across?

**Comment by tjwebb** [comment]
Yea; we've seen some other folks working on similar solutions. I created [sails-permissions](https://github.com/tjwebb/sails-permissions) to solve this problem in Waterline and not rely on any specific database features. sails-permissions was originally designed for a multi-tenant ERP system that ran on postgres. 

Partitioning as well as separate "physical" databases solve this problem in very similar ways.

**Comment by kheide** [comment]
So if I understand you correctly you recommend schema/partioning over separate databases?

Can you point me to where in 'sails-permission' the code for this is?

**Comment by tjwebb** [comment]
Partitioning is designed to distribute large tables across nodes for horizontal scalability. It's a performance optimization. I'm not seeing how it helps you achieve multi-tenancy; it solves a totally different kind of problem.

> So if I understand you correctly you recommend schema/partioning over separate databases?

No that's the opposite of what I'm recommending. The bullets in my previous comment are a list of problems with that approach. My recommendation is to implement multi-tenancy in the middleware layer using Waterline -- using sails-permissions as a starting point -- rather than in the database layer.

**Comment by kheide** [comment]
Gotcha! 

So this implementation would be more along the lines of what you're suggesting? https://github.com/balderdashy/sails/issues/1857#issuecomment-46210435

**Comment by tjwebb** [comment]
Yea something like that. The PR that resulted from that issue was somewhat more comprehensive than could be pulled into the sails core.

**Comment by sailsbot** [comment]
Thanks for posting, @kheide.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
